cpp/src/jni/dataset/proto/Types.proto

```diff
@@ -0,0 +1,149 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+syntax = "proto2";
+package types;
+
+option java_package = "org.apache.arrow.dataset";
```

Comment:
why is this file being introduced?

Reply:
Currently it is for passing datasets filter via JNI. I see Wes has recommended not to include filter bindings (in the bottom comment) in this version. I may consider removing it in a later commit.

Reply:
I think it's fine if you have a filter API in Java, but I think you should try to limit the exposure of the particular details of the current C++ API in Java. If the Java API has a 1-1 correspondence with the C++ API that's what I'm advising you against

Reply:
Thanks, as I knew some of the base design purposes of datasets API, dataframe API, and query engine system, so I think I would understand any of the concerns against a too detailed mapping. I'll try remove more JNI stuffs.

As for filter, If I understand correctly it's OK to keep Java API but the JNI mapping for filters is considered fragile, right? I can remove the mapping anyway but when users read parquet files from Java they'll not be able to filter row groups to reduce I/O. Which is extremely important when low-selectivity filters is specified.

Reply:
> As for filter, If I understand correctly it's OK to keep Java API but the JNI mapping for filters is considered fragile, right? I can remove the mapping anyway but when users read parquet files from Java they'll not be able to filter row groups to reduce I/O. Which is extremely important when low-selectivity filter is specified.

I'm advising you to define some kind of abstraction layer so that the _details_ of the C++ API are not exposed in the Java API. So if there are changes in the C++ API, then refactoring can take place only at the JNI C++ code and no changes will be required in Java.

Reply:
@wesm 

> So if there are changes in the C++ API, then refactoring can take place only at the JNI C++ code and no changes will be required in Java.

Sorry I am not quite getting it. Is it really needed to change Java code if c++ filter changed when it comes to the current PR? I mean, now the "layers" of the approach are something like:

```
  Java: DatasetTypes.Condition (ProtoBuf generated code)
                        |
               Java: byte[]
                        |
                 JNI Bridge
                        |
               C++: uint8_t*
                        |
   C++ types::Condition (ProtoBuf generated code)
                        |   <---- c++ function TranslateFilter(...)
   C++ arrow::dataset::Expression (Dataset Filters) 
```

The top 5 layers are all based on protobuf generated codes so we don't have to change any of them from Java-side when we want to make changes from C++. Even say if we want to change Java filter API we can also just modify file Types.proto. Or in future if we refactor C++ API arrow::dataset::Expression, we may have to make related changes to the function `TranslateFilter(...)` as well to make JNI things matched but we don't have to change Java codes anyway.

And the filter tree structures are not 1-1 mapped too, the definitions in Types.proto are much simpler than arrow::dataset::Expression so far, I didn't mean to make them strictly match. In future we can change C++ filters API without touching Java API and it is fine enough to me.

Reply:
I haven't reviewed the Java code, so it sounds like things are okay, I just wanted to warn you about the likely instability of the C++ API

Reply:
I see. At this time I'll try keeping them detached as far as possible.

Reply:
Can you start a thread on the mailing list about filter support.  I'm concerned with a proto specific to the java implementation of datasets.  Especially, because I think this essentially replicates what is in the Gandiva proto.  We should ideally have one representation for all the languages that are going to bind to the same functionality.

Reply:
Thanks for your suggestion, but they don't seem like the same functionality to me for dataset expression and Gandiva expression... In C++ code they are designed and placed differently and we'll face problems merging the JNI abstraction together. e.g. arrow::dataset::FieldExpression doesn't retain data type information (it is only represented by field name) where gandiva::FieldNode is type sensitive... I don't think it's a good idea to use any workaround to make them suitable for each other. What do you think?

Reply:
I do not have a solution off the top of my head.  But creating a new java->C++ protobuf abstraction seems like the the wrong way to go.  

As far as i can tell gandiva [protobuf declare column type optional](https://github.com/apache/arrow/blob/116672fed5cd8b7d80e5e8bd93615704aa10eca9/cpp/src/gandiva/proto/Types.proto#L93).  This needs more design which is why a discussion/proposal on the ML is warranted. 

It might make sense to split out the filtering components of this PR  so it doesn't get held up by the conversation.

Reply:
Yes I prefer removing them from this PR as I think of the implementation immature too. I can open ticket/discussion later.